File: mobile/lib/shared/models/message.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'message.freezed.dart';
part 'message.g.dart';

/// Message types for the messaging system
enum MessageType {
  @JsonValue('ground_admin_invitation')
  groundAdminInvitation,
  @JsonValue('ground_admin_application')
  groundAdminApplication,
  @JsonValue('ground_admin_approved')
  groundAdminApproved,
  @JsonValue('ground_admin_declined')
  groundAdminDeclined,
  @JsonValue('ground_admin_invitation_declined')
  groundAdminInvitationDeclined,
  @JsonValue('ground_admin_revocation')
  groundAdminRevocation,
  @JsonValue('ground_admin_stepdown_request')
  groundAdminStepdownRequest,
  @JsonValue('verify_new_issue')
  verifyNewIssue,
  @JsonValue('verify_fix')
  verifyFix,
  @JsonValue('member_registration')
  memberRegistration,
  @JsonValue('monthly_report')
  monthlyReport;

  String get displayName => switch (this) {
        groundAdminInvitation => 'Ground Admin Invitation',
        groundAdminApplication => 'Ground Admin Application',
        groundAdminApproved => 'Application Approved',
        groundAdminDeclined => 'Application Declined',
        groundAdminInvitationDeclined => 'Invitation Declined',
        groundAdminRevocation => 'Status Revoked',
        groundAdminStepdownRequest => 'Step Down Request',
        verifyNewIssue => 'Verify Issue',
        verifyFix => 'Verify Fix',
        memberRegistration => 'Member Registration',
        monthlyReport => 'Monthly Report',
      };
}

/// Message status
enum MessageStatus {
  unread,
  read,
  actioned,
  dismissed;

  String get displayName => switch (this) {
        unread => 'Unread',
        read => 'Read',
        actioned => 'Actioned',
        dismissed => 'Dismissed',
      };
}

/// A message in the user's inbox
@freezed
abstract class Message with _$Message {
  const Message._();

  const factory Message({
    required String id,
    required MessageType type,
    required String title,
    required String body,
    required String recipientId,
    required String recipientType,
    String? senderId,
    String? senderType,
    required MessageStatus status,
    String? actionType,
    String? relatedEntityId,
    String? relatedEntityType,
    String? actionResult,
    Map<String, dynamic>? metadata,
    required DateTime createdAt,
    DateTime? readAt,
    DateTime? actionedAt,
    DateTime? expiresAt,
  }) = _Message;

  /// Whether this message has pending actions
  bool get hasActions => actionType != null && status != MessageStatus.actioned;

  /// Whether this message is unread
  bool get isUnread => status == MessageStatus.unread;

  /// Whether this message is related to verification
  bool get isVerificationMessage =>
      type == MessageType.verifyNewIssue || type == MessageType.verifyFix;

  /// Whether this message is related to ground admin flow
  bool get isGroundAdminMessage =>
      type == MessageType.groundAdminInvitation ||
      type == MessageType.groundAdminApplication ||
      type == MessageType.groundAdminApproved ||
      type == MessageType.groundAdminDeclined ||
      type == MessageType.groundAdminInvitationDeclined ||
      type == MessageType.groundAdminRevocation ||
      type == MessageType.groundAdminStepdownRequest;

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);
}

/// Paginated response for messages
@freezed
abstract class MessageListResponse with _$MessageListResponse {
  const factory MessageListResponse({
    required List<Message> items,
    required int total,
    required int page,
    required int unreadCount,
  }) = _MessageListResponse;

  factory MessageListResponse.fromJson(Map<String, dynamic> json) =>
      _$MessageListResponseFromJson(json);
}

/// Request to perform an action on a message
@freezed
abstract class MessageActionRequest with _$MessageActionRequest {
  const factory MessageActionRequest({
    required String action,
    String? note,
  }) = _MessageActionRequest;

  factory MessageActionRequest.fromJson(Map<String, dynamic> json) =>
      _$MessageActionRequestFromJson(json);
}

File: web/src/shared/types/message.ts
/**
 * Message types for the platform messaging system
 */
export type MessageType =
  | 'ground_admin_invitation'
  | 'ground_admin_application'
  | 'ground_admin_approved'
  | 'ground_admin_declined'
  | 'ground_admin_invitation_declined'
  | 'ground_admin_invitation_accepted'
  | 'ground_admin_revocation'
  | 'ground_admin_stepdown_request'
  | 'verify_new_issue'
  | 'verify_fix'
  | 'member_registration'
  | 'monthly_report';

/**
 * Status of a message in the system
 */
export type MessageStatus = 'unread' | 'read' | 'actioned' | 'dismissed';

/**
 * Recipient type for messages
 */
export type RecipientType = 'member' | 'admin';

/**
 * Sender type for messages
 */
export type SenderType = 'member' | 'admin' | 'system';

/**
 * Message entity representing platform communications
 */
export interface Message {
  id: string;
  type: MessageType;
  title: string;
  body: string;
  recipientId: string;
  recipientType: RecipientType;
  senderId?: string;
  senderType?: SenderType;
  status: MessageStatus;
  actionType?: string;
  relatedEntityId?: string;
  relatedEntityType?: string;
  actionResult?: string;
  metadata?: Record<string, unknown>;
  createdAt: string;
  readAt?: string;
  actionedAt?: string;
  expiresAt?: string;
}

/**
 * Response from messages list endpoint
 */
export interface MessageListResponse {
  items: Message[];
  total: number;
  page: number;
  unreadCount: number;
}

/**
 * Request to perform an action on a message
 */
export interface MessageActionRequest {
  action: string;
  note?: string;
}

/**
 * Parameters for filtering messages
 */
export interface MessageFilterParams {
  status?: MessageStatus;
  type?: MessageType;
  page?: number;
  size?: number;
}

/**
 * Action types that messages can have
 */
export const MESSAGE_ACTION_TYPES = {
  ACCEPT_DECLINE: 'accept_decline',
  APPROVE_REJECT: 'approve_reject',
  CONFIRM_VERIFY: 'confirm_verify',
  ACKNOWLEDGE: 'acknowledge',
  VIEW: 'view',
} as const;

/**
 * Map message types to their i18n label keys
 */
export const MESSAGE_TYPE_LABELS: Record<MessageType, string> = {
  ground_admin_invitation: 'messages.types.groundAdminInvitation',
  ground_admin_application: 'messages.types.groundAdminApplication',
  ground_admin_approved: 'messages.types.groundAdminApproved',
  ground_admin_declined: 'messages.types.groundAdminDeclined',
  ground_admin_invitation_declined: 'messages.types.groundAdminInvitationDeclined',
  ground_admin_invitation_accepted: 'messages.types.groundAdminInvitationAccepted',
  ground_admin_revocation: 'messages.types.groundAdminRevocation',
  ground_admin_stepdown_request: 'messages.types.groundAdminStepdownRequest',
  verify_new_issue: 'messages.types.verifyNewIssue',
  verify_fix: 'messages.types.verifyFix',
  member_registration: 'messages.types.memberRegistration',
  monthly_report: 'messages.types.monthlyReport',
};
